### ISSUE TITLE
Replace unmaintained `isbnlib` dependency

### ISSUE THREAD

**Issue by mycorrhizan** [papis/papis#948]
No commits have been made since 2024-07-28 ([ddc62f2](https://github.com/xlcnd/isbnlib/commit/ddc62f2b77214205a7413b5d049c2a008c5271d4)) and CI/CD tests have been consistently failing since 2023-10-19 ([#2007](https://github.com/xlcnd/isbnlib/actions/runs/6570266426)) starting with the Windows tests, with the latest CI/CD test as of this issue failing for all tests ([#2404](https://github.com/xlcnd/isbnlib/actions/runs/11876293124)).

**Comment by alexfikl** [comment]
Do you know of any possible replacement?

From a quick skim through the failures, they seem to be some HTTP failures for too many requests, so not quite that dire, but they might hide other things :( The contributions graph looks scary https://github.com/xlcnd/isbnlib/graphs/contributors

**Comment by alexfikl** [comment]
> The most suitable replacement I could find is [idutils](https://pypi.org/project/idutils/), though I'm unsure if it covers all requirements. At the very least it does seem to be consistently maintained.

The feature we actually use from isbnlib is downloading metadata from various sources based on the ISBN. I think they support Google Books and OpenLibrary.org (plus a bunch of plugins), so that would be hard to replace :\

**Comment by alexfikl** [comment]
> Would this be a suitable replacement for those sources?

To start with, we can definitely add some alternative to `isbnlib` for working with ISBNs using those libraries.

**Comment by alexfikl** [comment]
> Sounds like a plan. Is there a list of available sources in the documentation or source code I can refer to for finding replacements?

I meant that we can add some importer plugins as alternatives to the `isbnlib` ones. There's some documentation on how to add those here
https://papis.readthedocs.io/en/latest/plugins.html
and a lot of examples through the code (all the classes that inherit from `papis.importer.Importer`).

**Comment by fdf-uni** [comment]
Hey, I recently stumbled upon this issue as I also got the warning from papis regarding `setuptools`.
I've since implemented a small (~200 lines) python script which I think would cover the needs of papis (i.e. checking whether a string is a valid ISBN and fetching metadata from Google Books, Open Library and Wikipedia), at least from what I've gathered after grepping the code for `isbnlib`.

I'd happily contribute it in some way if you'd wish so and could perhaps give some instructions in what way to do this. :)
The only problem however might be that I had a look at some parts of `isbnlib`, so I'm not sure how this would work with licensing (I've rewritten most parts, but copied some API endpoints and such; in case this would be preferred, I could also write down a "spec" or something).

**Comment by alexfikl** [comment]
> I'd happily contribute it in some way if you'd wish so and could perhaps give some instructions in what way to do this. :)

You can just add an `Importer` plugin that more or less mirrors the `isbnlib` one. For instructions:

* There's a bit of a wordy explanation of how the `Downloader` plugins work (the `Importer` ones are just simpler) over here: https://papis.readthedocs.io/en/latest/plugins.html#importer.
* The API documentation for an importer is found here: https://papis.readthedocs.io/en/latest/developer_reference.html#module-papis.importer

In short, you just need to make a subclass of `papis.importer.Importer` and implement the `match` and `fetch_data` functions. After that's done, you can add it as a plugin to `pyproject.toml`. You can make a little PR with whatever in progress code you have and we can go from there :grin:

That being said, it would be nice if some external library did the ISBN fetching for us. Seeing how `isbnlib` is going to break soon enough (with `pkg_resources` going away), then maintaining 200-ish lines of code until something comes up shouldn't be a big burden..

> so I'm not sure how this would work with licensing

Papis is licensed as `GPL-3.0-or-later` and `isbnlib` is `LGPL-3.0-or-later`, so I don't think there should be any problem there even if you copy-pasted bigger chunks. You'll just need to add a little attribution blurb somewhere.

**Comment by hseg** [comment]
Stumbled into https://github.com/ImranR98/ISBN-Lookup, which led me to look for other libraries we could wrap (I was even toying with using FFI). Posting my findings here in case they're of use:

<details><summary>Only relevant as prior art</summary>
- https://github.com/ImranR98/ISBN-Lookup (js->python is much too hard, project is too young to be reliable)
- https://github.com/CuriousCorrelation/recon_metadata (project seems abandoned)
- https://github.com/NYPL/libbib (not sure how maintained it is, but WorldCat is a source I'm not sure `isbnlib` ever had)
</details>

<details><summary>Abandoned, but could be adopted</summary>
- https://github.com/xlcnd/isbnlib (the OG)
</details>

<details><summary>Maintained, potentially problematic</summary>
- https://github.com/HzaCode/OneCite (impedance mismatch re usecases, only uses Google Books, might be overly clever)
</details>

**Comment by alexfikl** [comment]
For what it's worth, `isbnlib` did have WorldCat support through xISBN, but it was removed in https://github.com/xlcnd/isbnlib/commit/664a7cb9b75bf8adcd4610f564e134a88671e1a7 when the xISBN service went down. The newer OCLC backend requires an API key that they don't seem to give out very freely (i.e. mostly to actual libraries and institutions).

**Comment by hseg** [comment]
Head's up -- someone has forked `isbnlib` and is slowly restarting development under the name `isbnlib2`. Changes seem OK by my sniff tests, but you may want to check for yourself too.
https://github.com/hans-fritz-pommes/isbnlib

**Comment by hseg** [comment]
That fork also eg merges in xlcnd/isbnlib#145 in https://github.com/hans-fritz-pommes/isbnlib/commit/7e8e0101f8884c5fb9c987387e6d39723841adaf, so our `setuptools` version bound would be removable.

**Comment by alexfikl** [comment]
> Head's up -- someone has forked isbnlib and is slowly restarting development under the name isbnlib2.

The API seems to be compatible for now, so that should just work, right? It doesn't seem like they can be installed side-by-side though. Would we just need to add it to the `optional` group? Not quite sure..

>  so our setuptools dependency would be able to be removed.

I've made all these optional in #1147, so it's less of an issue.

**Comment by hseg** [comment]
> > Head's up -- someone has forked isbnlib and is slowly restarting development under the name isbnlib2.
>
> The API seems to be compatible for now, so that should just work, right? It doesn't seem like they can be installed side-by-side though. Would we just need to add it to the *optional* group? Not quite sure..

Afaict, they haven't broken api compatibility yet, but that needs testing. Re side-by-side installation - yes, it seems `isbnlib2`'s packaging choices make it collide with `isbnlib`. In fact, checking PyPI procedures for taking over development of abandoned projects (PEP 541), it seems `isbnlib2` should just file an issue to inherit the old `isbnlib` name. @hans-fritz-pommes, have you done this?
We might wait for that part of the situation to stabilize, or list `isbnlib2` as an alternative to `isbnlib`. Mainly wanted to bring its existence to your attention.

> so our setuptools dependency would be able to be removed.
>
> I've made all these optional in #1147, so it's less of an issue.

True - but it's still a wart that could be removed.
Also, I had somehow missed that the commits were in a PR, so my commit message for the AUR edit for that is lacking that pointer. Made a new empty commit to backlink to it.

**Comment by hans-fritz-pommes** [comment]
Hello,
I didn't do nothing yet to inherit the name.
But why not
I will have a look.

- I opened a pypi-issue to overtake the project

**Comment by alexfikl** [comment]
We went with copying over a small part of `isbnlib` that we actually use, so this should be solved. Maybe we'll move to `isbnlib2` in the future (if the code turns out to be annoying to maintain), but that shouldn't be a breaking change :grin:

xref: #1102

**Comment by hseg** [comment]
That's great, thanks for the heads' up! Will try to have an update out by the weekend

**Comment by hseg** [comment]
This fell off my radar, as I was swamped in the last month.

Was just about to release this, when I noticed that in a chroot I get a failure on `tests/test_utils.py::test_yaml_unicode_dump` and on my local system I get a failure on `tests/commands/test_default.py::test_no_config_shows_init_hint`. Investigating...

**Comment by alexfikl** [comment]
> Was just about to release this, when I noticed that in a chroot I get a failure on `tests/test_utils.py::test_yaml_unicode_dump` and on my local system I get a failure on `tests/commands/test_default.py::test_no_config_shows_init_hint`. Investigating...

Those don't sound like they're in any way related to the `isbnlib` changes?

**Comment by hseg** [comment]
No, but I block releases on the expectation that they test cleanly. Although, I perhaps should revise that policy, especially for a rolling-release package like `papis-git`.
EDIT: Accepted your reasoning, version pushed.
